Selenium captchas (and lots of other stuff)

stefanct opened the pull request:
Among other things this tackles #131 and #137. There are some changes that might be controversial, most notably "Do not sleep between page requests". Arguably the most important patch is "Fix the fix for parsing venues and years " because it fixes something I broke in the last PR (sorry).

I did not add extra tests to the script but run the existing tests in test_module.py on Linux without failures (apart from the unclosed sockets warnings) and I have been testing these changes quite a bit with and without an external proxy generator based on `proxyscrape` (not to be confused with what's currently worked on by Panos Ipeirotis.

stefanct commented:
I am trying to convince the CI script to install selenium in case you wonder... ;)

stefanct commented:
OK, that's no longer the culprit for failed CI tests. However, I am not sure how CI should work with selenium/captcha handling enabled anyway. I'll fixup the selenium commit to incorporate the CI changes and will then wait for feedback.

stefanct commented:
Gnarrrr, just found another possibility to break the venue/year parsing. There is a fourth way GS presents the data, e.g., `NR Satish, PY Droz - Technical Re port UCB/ERL M - academia.edu`, i.e. just the "venue" without a year. Working on a fix that tests for four numerals...

stefanct commented:
And *of course* there is also `NRC de Sousa - Departamento de Informtica, Universidade do Minho - gec.di.uminho.pt`...

stefanct commented:
OK, my ~1000 natural test cases and the manually built one check out. Hope that finally fixes the venue/year parsing for good.

ipeirotis commented:
Add it in requirements.txt :)

On Fri, Jun 5, 2020 at 17:43 Stefan Tauner <notifications@github.com> wrote:

> I am trying to convince the CI script to install selenium in case you
> wonder... ;)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scholarly-python-package/scholarly/pull/143#issuecomment-639837854>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFH4KVQCV7XBUFXWQGDWDRVFRHFANCNFSM4NVGUMDA>
> .
>
-- 
Sent from Mobile

silvavn commented on scholarly/publication.py:
```diff
@@ -254,7 +254,7 @@ def fill(self):
     @property
     def citedby(self) -> _SearchScholarIterator or list:
-        """Searches GScholar for other articles that cite this Publication and
+        """Searches Google Scholar for other articles that cite this Publication and
```
This was to make sure that the code is PEP-8 Compliant.
Is it still?

stefanct replied:
Hm, I am not sure exactly how this is related to PEP-8 since I perceived that as a human-readable comment only. Can you please be more specific?

silvavn commented on scholarly/_scholarly.py:
```diff
@@ -23,12 +23,12 @@ def set_retries(self, num_retries: int):
         :param http: the http proxy address
         :type http: str
-        :param https: the https proxy address
+        :param https: the https proxy (default to the same as http)
```
Is this really true? the code presented in this commit just sets it to none and makes it optional

stefanct replied:
Yes, see cddbbcd3adfa06df8a35eff24357bc99f206409b, which is the "previous change" I mentioned in the commit message. I could make it more explicit if you like.

silvavn commented:
Overall this is very good work. Check the selenium branch that I have created. There is code ready to support both chrome and firefox there. If you could integrate it to your push request that would be awesome!

I know this is controversial, but we should really start slowing down the requests and let advanced users overcome those barriers. My reasoning is: if we release the package without any timeout everybody using it will be basically doing a lot of requests at the same time. That behaviour is very similar to a DDoS attack, which is concerning.

About the PEP-8 specification, you can use a code linter, or read the warnings that are raised in the CI framework here on github (e.g. more than 79 columns )

ipeirotis commented:
> Overall this is very good work. Check the selenium branch that I have created. There is code ready to support both chrome and firefox there. If you could integrate it to your push request that would be awesome!
> 
> I know this is controversial, but we should really start slowing down the requests and let advanced users overcome those barriers. My reasoning is: if we release the package without any timeout everybody using it will be basically doing a lot of requests at the same time. That behaviour is very similar to a DDoS attack, which is concerning.
> 
> About the PEP-8 specification, you can use a code linter, or read the warnings that are raised in the CI framework here on github (e.g. more than 79 columns )

For the TIMEOUT, notice that this is not for waiting between requests, it is to wait for the network to respond. Leaving it infinite is bad, as the code may hang, but making it too low can cause unnecessary repeated requests. I would still keep it at 10 seconds.

For PEP8 compliance, we can insist on Black as the code formatter. It is very opinionated, which ends up being good for team projects: no individual opinions on how the code should be formatted :-)

Finally, for the refactoring: In the `proxy-generator` branch, I am taking out all the proxy/tor handling. The networking component needs to only call "get_next_proxy()", set the new proxy, and retry. Hopefully, I will manage to finish that component together with the tests in the next couple of days.

stefanct commented:
[I have rearranged the quotes to better fit the discussed topics]

> > About the PEP-8 specification, you can use a code linter, or read the warnings that are raised in the CI framework here on github (e.g. more than 79 columns )

> For PEP8 compliance, we can insist on Black as the code formatter. It is very opinionated, which ends up being good for team projects: no individual opinions on how the code should be formatted :-)

Ah, it was about line lengths, OK. That should not influence the spelling of words though. ;)
While I personally have very strong opinions against algorithmically enforced styles (since they *always* result in ridiculous and actually harmful formatting in some corner cases), I abstain from trying to convince you otherwise. I do not intend to continuously contribute to this project - I just needed to get it to work for me and couldn't wait. How can I run the PEP-8 checks locally to easily test and fix my patches?

> > I know this is controversial, but we should really start slowing down the requests and let advanced users overcome those barriers. My reasoning is: if we release the package without any timeout everybody using it will be basically doing a lot of requests at the same time. That behaviour is very similar to a DDoS attack, which is concerning.

> For the TIMEOUT, notice that this is not for waiting between requests, it is to wait for the network to respond. Leaving it infinite is bad, as the code may hang, but making it too low can cause unnecessary repeated requests. I would still keep it at 10 seconds.

This discussion possibly mixes two things: timeout handling via `Navigator`'s `_TIMEOUT` and the `sleep` call in `_get_page` that I removed (in separate commits).

In general we are talking mainly about Google servers here that is protected by captchas (I don't deem the proxies or even Tor as relevant regarding DOSing since they are essentially network infrastructure with the sole purpose of transporting packets) and access we them in a single thread without pipelining. There is no way this even comes close to something that raises their eyebrows IMHO as long as there aren't 1000s of concurrent scholarly users. But again, I have no stakes in this project and you bear the responsibility.

However, let me explain why I refined the `_TIMEOUT` handling: While implementing/testing my proxy generator and gotcha handling it became clear that the long timeouts are tremendously slowing down `_check_proxy` for proxies that do not work properly (which is the vast majority of free proxies in my experience). Effectively this *was* determining the rate of proxy tests and eventually constituting the overall runtime of my queries (for the metadata, i.e. without `fill` calls, of 1000 papers/100 search result pages).

Thus this is a major performance bottleneck when using unreliable proxies. It is a big difference between having to wait for 10*10 seconds to find a working proxy than 10*3 seconds obviously. Overall I found that the difference in proxy test time is bigger than the time my actual queries ran(!) - with my other changes applied of course.

One possible solution would be to use a shorter timeout for the first check than for later accesses. I refrained from doing so since I don't see the benefit of requiring a shorter initial reply (although the replies by the test servers should be faster than Scholar's reply... since the latter is Google that's not necessarily the case - especially with badly chosen test pages like the enormous twitter front page that was set previously - that's one reason why I changed that too, to the smallest and potentially quickest I could easily find). As you say the initial limit might be too low for otherwise working proxies and thus I added the adapting algorithm to increase it on timeout errors once it worked initially.

I think that's a good approach for the environment I have in mind with very little disadvantages when used in others (e.g., high-quality payed proxies or w/o proxies/Tor at all).


The other thing is the removal of the sleep calls that were meant to enforce a minimum time between requests ("Space a bit the requests to avoid overloading the servers"). I think this is what the DOS comment was actually about (github didn't show the context/related patch of this comment but appended it to the whole PR). While this seems to make sense the values stem from gut feeling I presume.

However, Google gives us some hints to what they seem to care about and that seems to be rather the *number* of requests in a rather large amount of time not so much the time *between* them. At least that's what I was experiencing when testing: it does not really matter how fast you fetch the result pages but how many of them within some "session length" amount of time.
My conclusion is that they don't care at all about getting "DOSed" but simply don't want to get their database scraped to larger extents. And from that follows that making scholarly slower has no advantage whatsoever and simply makes user's lives more miserable.

I am fine with you leaving the sleep in or implementing any other DOS avoidance scheme - it's totally your choice and I'll stay out of further discussions on the topic.


> > Overall this is very good work. Check the selenium branch that I have created. There is code ready to support both chrome and firefox there. If you could integrate it to your push request that would be awesome!

> Finally, for the refactoring: In the `proxy-generator` branch, I am taking out all the proxy/tor handling. The networking component needs to only call "get_next_proxy()", set the new proxy, and retry. Hopefully, I will manage to finish that component together with the tests in the next couple of days.

I have seen the proxy work. It's unfortunate that we were working on the same stuff in parallel. :( I have to be honest: I won't have the time to incorporate your proxy changes/base mine on top of them - they looked quite extensive. I have already spent way more time on this than I can argue in good conscience unfortunately. I tried my best to make the patch set easy to refine and pick the important parts while ignoring the controversial change so that you can decide. Please don't expect me to rework the patch set considerably. I'll take a look at the chrome support though (and I can/will fix smaller stuff like the PEP thing if you point them out of course).

stefanct commented:
@silvavn i cant find the repo/branch with the chrome support. https://github.com/scholarly-python-package/scholarly/commits/selenium can't be what you mean AFAICT

silvavn commented:
@stefanct I coded it and forgot to push 😆. Most of it is there now.
